src/test/java/org/humancellatlas/ingest/IngestStateTrackingApplicationTests.java

```diff
@@ -251,6 +251,123 @@ public void testSuccessfulEventRunthroughWithoutArchiving() {
         assertEquals(SubmissionState.COMPLETE, state);
     }
 
+    @Test
+    public void testSuccessfulEventRunthroughAfterUpdate() {
+        assertTrue(submissionStateMonitor.isMonitoring(envelopeRef));
+        SubmissionState state = submissionStateMonitor.findCurrentState(envelopeRef);
+        assertEquals(SubmissionState.PENDING, state);
+
+        log.debug("Adding content");
+        assertTrue(submissionStateMonitor.notifyOfMetadataDocumentState(documentRef, envelopeRef, MetadataDocumentState.DRAFT));
+        state = submissionStateMonitor.findCurrentState(envelopeRef);
+        assertEquals(SubmissionState.DRAFT, state);
+
+        submissionStateMonitor.notifyOfMetadataDocumentState(documentRef, envelopeRef, MetadataDocumentState.VALIDATING);
+        state = submissionStateMonitor.findCurrentState(envelopeRef);
+        assertEquals(SubmissionState.VALIDATING, state);
+
+        // wait for a bit to simulate validation happening
+        try {
+            TimeUnit.SECONDS.sleep(10);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+
+        submissionStateMonitor.notifyOfMetadataDocumentState(documentRef, envelopeRef, MetadataDocumentState.VALID);
+        // wait for a bit to allow propagation of cascade events
+        try {
+            TimeUnit.SECONDS.sleep(1);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+
+        state = submissionStateMonitor.findCurrentState(envelopeRef);
+        assertEquals(SubmissionState.VALID, state);
+
+        // assert 0 documents in extended state map
+        Map<String, MetadataDocumentState> metadataDocumentStateMap = (Map<String, MetadataDocumentState>) submissionStateMonitor.findStateMachine(UUID.fromString(envelopeRef.getUuid())).get().getExtendedState().getVariables().get(Constants.METADATA_DOCUMENT_TRACKER);
```

Comment:
This seems like a test that knows about the dependencies of dependencies (and even further). In other words, it violates Law of Demeter. Shouldn't this be a method of StateMachine (or another direct dependency)?

Reply:
I more or less copied the logic from [this test](https://github.com/ebi-ait/ingest-state-tracking/blob/9401524b5b48630c2b40e104d72db19c8e5f2721/src/test/java/org/humancellatlas/ingest/IngestStateTrackingApplicationTests.java#L85). I agree with you though, it's a long chain. I'm not sure how I would refactor this though. The relationship between `MetadataDocument`, its state, and state machine is not clear to me